File: examples/rest-api/producer_pal.py
#!/usr/bin/env python3

"""Producer Pal REST API example (Python, no dependencies).

Usage: python producer_pal.py
"""

import json
import sys
import urllib.error
import urllib.parse
import urllib.request

BASE_URL = "http://localhost:3350"


def list_tools():
    """List all available tools."""
    req = urllib.request.Request(f"{BASE_URL}/api/tools")
    with urllib.request.urlopen(req) as res:
        return json.loads(res.read())["tools"]


def call_tool(name, args=None, *, format=None, timeout_ms=None):
    """Call a Producer Pal tool by name.

    Args:
        name:       Tool name (e.g. "ppal-read-live-set").
        args:       Tool input dict. Defaults to {}.
        format:     "json" or "compact". Default "compact" returns `result`
                    as a token-efficient JS-literal string. "json" returns
                    `result` as a parsed value (dict/list/etc.) and surfaces
                    any warnings as a separate `warnings` list.
        timeout_ms: Per-request timeout override (1-60000).
    """
    params = {}
    if format:
        params["format"] = format
    if timeout_ms is not None:
        params["timeoutMs"] = str(timeout_ms)

    url = f"{BASE_URL}/api/tools/{name}"
    if params:
        url = f"{url}?{urllib.parse.urlencode(params)}"

    data = json.dumps(args or {}).encode()
    req = urllib.request.Request(
        url,
        data=data,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req) as res:
        return json.loads(res.read())


def main():
    # List available tools
    print("Available tools:")
    tools = list_tools()
    for tool in tools:
        print(f"  {tool['name']} - {tool['description'][:60]}...")

    # Inspect a tool's input schema
    read_track = next(t for t in tools if t["name"] == "ppal-read-track")
    print(f"\nSchema for {read_track['name']}:")
    print(json.dumps(read_track["inputSchema"], indent=2))

    # Read tracks in the Live Set (default compact format)
    print("\nReading tracks (compact format):")
    tracks = call_tool("ppal-read-live-set")
    print(tracks["result"])

    # Same call with format="json" — result is a parsed dict you can index into
    print("\nReading tracks (parsed JSON):")
    tracks_json = call_tool("ppal-read-live-set", format="json")
    print(f"tempo: {tracks_json['result']['tempo']}")
    print(f"scale: {tracks_json['result']['scale']}")

    # Override timeout for a potentially long call
    print("\nReading track 0 with all clips (10s timeout):")
    track = call_tool(
        "ppal-read-track",
        {"trackIndex": 0, "include": ["session-clips", "arrangement-clips"]},
        timeout_ms=10000,
    )
    print(track["result"])


if __name__ == "__main__":
    try:
        main()
    except urllib.error.URLError as e:
        if "Connection refused" in str(e.reason):
            print(
                "Could not connect to Producer Pal."
                " Is Ableton Live running with the Producer Pal device?",
                file=sys.stderr,
            )
        else:
            raise
        sys.exit(1)

File: examples/rest-api/producer-pal.mjs
#!/usr/bin/env node

// Producer Pal REST API client (Node.js, no dependencies)
// Requires Node.js 18+ for built-in fetch
//
// Usage:
//   node producer-pal.mjs <tool-name> [json-args] [options]
//   node producer-pal.mjs --list-tools [options]
//
// Options:
//   --url URL              REST API base URL (default: http://localhost:3350)
//   --format json|compact  Response format (default: json)
//   --timeout-ms N         Per-call timeout in milliseconds
//   --help, -h             Show help
//
// Examples:
//   node producer-pal.mjs --list-tools
//   node producer-pal.mjs ppal-read-live-set
//   node producer-pal.mjs ppal-read-track '{"trackIndex": 0}'
//   node producer-pal.mjs --list-tools | jq -r '.tools[].name'
//   node producer-pal.mjs ppal-read-live-set | jq .result.tempo

function usage() {
  return [
    "Usage:",
    "  node producer-pal.mjs <tool-name> [json-args] [options]",
    "  node producer-pal.mjs --list-tools [options]",
    "",
    "Options:",
    "  --url URL              REST API base URL (default: http://localhost:3350)",
    "  --format json|compact  Response format (default: json)",
    "  --timeout-ms N         Per-call timeout in milliseconds",
    "  --help, -h             Show help",
    "",
    "Examples:",
    "  node producer-pal.mjs --list-tools",
    "  node producer-pal.mjs ppal-read-live-set",
    "  node producer-pal.mjs ppal-read-track '{\"trackIndex\": 0}'",
  ].join("\n");
}

function parseArgs(argv) {
  const opts = {
    url: "http://localhost:3350",
    format: "json",
    timeoutMs: undefined,
    listTools: false,
    toolName: undefined,
    toolArgs: {},
  };
  const positional = [];

  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    const takeValue = () => {
      const v = argv[++i];
      if (v === undefined) throw new Error(`Missing value for ${arg}`);
      return v;
    };

    if (arg === "--url") opts.url = takeValue();
    else if (arg === "--format") opts.format = takeValue();
    else if (arg === "--timeout-ms") opts.timeoutMs = Number(takeValue());
    else if (arg === "--list-tools") opts.listTools = true;
    else if (arg === "--help" || arg === "-h") {
      console.log(usage());
      process.exit(0);
    } else if (arg.startsWith("--")) {
      throw new Error(`Unknown argument: ${arg}`);
    } else {
      positional.push(arg);
    }
  }

  if (opts.format !== "json" && opts.format !== "compact") {
    throw new Error(
      `Invalid --format: ${opts.format} (must be 'json' or 'compact')`,
    );
  }

  if (!opts.listTools) {
    if (positional.length === 0) {
      throw new Error(
        "Missing tool name. Use --list-tools to discover tools, or pass a tool name as the first argument.",
      );
    }
    opts.toolName = positional[0];
    if (positional.length > 1) {
      try {
        opts.toolArgs = JSON.parse(positional[1]);
      } catch (err) {
        throw new Error(`Invalid JSON for tool args: ${err.message}`);
      }
    }
    if (positional.length > 2) {
      throw new Error(`Unexpected positional argument: ${positional[2]}`);
    }
  }

  return opts;
}

/**
 * GET /api/tools — returns the full envelope `{tools: [...]}` as parsed JSON.
 * The tool list endpoint always returns JSON; it has no `?format` toggle.
 */
async function listTools(baseUrl) {
  const res = await fetch(`${baseUrl}/api/tools`);
  if (!res.ok) throw new Error(`HTTP ${res.status}: ${await res.text()}`);
  return res.json();
}

/**
 * Call a Producer Pal tool by name with the given args.
 *
 * In json format (the default), `result` is a parsed value (object/array/etc.)
 * and any warnings are surfaced as a separate `warnings: string[]` field. In
 * compact format, `result` is a token-efficient JS-literal string with
 * warnings inlined as `WARNING: ...` lines.
 */
async function callTool(baseUrl, name, args, { format, timeoutMs } = {}) {
  const params = new URLSearchParams({ format });
  if (timeoutMs != null) params.set("timeoutMs", String(timeoutMs));

  const res = await fetch(`${baseUrl}/api/tools/${name}?${params}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(args),
  });
  if (!res.ok) throw new Error(`HTTP ${res.status}: ${await res.text()}`);
  return res.json();
}

async function main() {
  const cli = parseArgs(process.argv.slice(2));

  if (cli.listTools) {
    console.log(JSON.stringify(await listTools(cli.url), null, 2));
    return;
  }

  const response = await callTool(cli.url, cli.toolName, cli.toolArgs, {
    format: cli.format,
    timeoutMs: cli.timeoutMs,
  });
  if (response.isError) {
    console.error("API error:", response.result);
    process.exit(1);
  }
  console.log(JSON.stringify(response, null, 2));
}

main().catch((err) => {
  if (err.cause?.code === "ECONNREFUSED") {
    console.error(
      "Could not connect to Producer Pal. Is Ableton Live running with the Producer Pal device?",
    );
  } else {
    console.error(err.message ?? err);
  }
  process.exit(1);
});
